File: templates/py/protocols/mavlink_link.py
from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator
from typing import Protocol, runtime_checkable


@runtime_checkable
class MavlinkLink(Protocol):
    """Transport-agnostic MAVLink byte stream."""

    async def send(self, data: bytes) -> None:
        """Send raw MAVLink frame bytes to the remote peer."""
        ...

    def receive(self) -> AsyncIterator[bytes]:
        """Incoming raw bytes from the remote peer."""
        ...

    async def close(self) -> None:
        """Release link resources. Default implementation is a no-op."""
        ...


class _VirtualMavlinkEndpoint:
    def __init__(self, bus: VirtualMavlinkBus) -> None:
        self._bus = bus
        self._queue: asyncio.Queue[bytes | None] = asyncio.Queue()
        self._closed = False

    async def send(self, data: bytes) -> None:
        if self._closed:
            raise RuntimeError("VirtualMavlinkEndpoint is closed")
        self._bus._deliver(bytes(data), self)

    async def receive(self) -> AsyncIterator[bytes]:
        while not self._closed:
            data = await self._queue.get()
            if data is None:
                break
            yield data

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        await self._queue.put(None)
        self._bus._endpoints.remove(self)


class VirtualMavlinkBus:
    """In-memory link for tests and virtual examples."""

    def __init__(self) -> None:
        self._endpoints: list[_VirtualMavlinkEndpoint] = []

    def create_endpoint(self) -> MavlinkLink:
        endpoint = _VirtualMavlinkEndpoint(self)
        self._endpoints.append(endpoint)
        return endpoint

    def _deliver(self, data: bytes, sender: _VirtualMavlinkEndpoint) -> None:
        for endpoint in self._endpoints:
            if endpoint is not sender and not endpoint._closed:
                endpoint._queue.put_nowait(data)

    async def close_all(self) -> None:
        endpoints = list(self._endpoints)
        for endpoint in endpoints:
            await endpoint.close()

File: templates/js/protocols/mavlink_link.js
/** Transport-agnostic MAVLink byte stream. */

/**
 * @typedef {object} MavlinkLink
 * @property {(data: Uint8Array) => Promise<void>} send
 * @property {AsyncIterable<Uint8Array>} receive
 * @property {() => Promise<void>} [close]
 */

class VirtualMavlinkEndpoint {
  constructor(bus) {
    this._bus = bus;
    this._queue = [];
    this._waiters = [];
    this._closed = false;

    this.receive = {
      [Symbol.asyncIterator]: () => ({
        next: () => {
          if (this._closed && this._queue.length === 0) {
            return Promise.resolve({ done: true, value: undefined });
          }
          const queued = this._queue.shift();
          if (queued !== undefined) {
            return Promise.resolve({ done: false, value: queued });
          }
          return new Promise((resolve, reject) => {
            this._waiters.push({ resolve, reject });
          });
        },
      }),
    };
  }

  async send(data) {
    if (this._closed) {
      throw new Error('VirtualMavlinkEndpoint is closed');
    }
    this._bus._deliver(new Uint8Array(data), this);
  }

  async close() {
    if (this._closed) {
      return;
    }
    this._closed = true;
    while (this._waiters.length > 0) {
      const waiter = this._waiters.shift();
      waiter.resolve({ done: true, value: undefined });
    }
    this._bus._removeEndpoint(this);
  }

  _emit(data) {
    if (this._closed) {
      return;
    }
    const waiter = this._waiters.shift();
    if (waiter !== undefined) {
      waiter.resolve({ done: false, value: data });
      return;
    }
    this._queue.push(data);
  }
}

/** In-memory link for tests and virtual examples. */
export class VirtualMavlinkBus {
  constructor() {
    this._endpoints = [];
  }

  /** Create a new endpoint on this bus. */
  createEndpoint() {
    const endpoint = new VirtualMavlinkEndpoint(this);
    this._endpoints.push(endpoint);
    return endpoint;
  }

  _deliver(data, sender) {
    for (const endpoint of this._endpoints) {
      if (endpoint !== sender) {
        endpoint._emit(data);
      }
    }
  }

  _removeEndpoint(endpoint) {
    const index = this._endpoints.indexOf(endpoint);
    if (index >= 0) {
      this._endpoints.splice(index, 1);
    }
  }

  /** Close every endpoint on the bus. */
  async closeAll() {
    const endpoints = [...this._endpoints];
    for (const endpoint of endpoints) {
      await endpoint.close();
    }
  }
}

/** Simple pub/sub stream for protocol-layer events (no external deps). */
export class EventStream {
  constructor() {
    this._listeners = new Set();
  }

  subscribe(listener) {
    this._listeners.add(listener);
    return () => this._listeners.delete(listener);
  }

  emit(value) {
    for (const listener of [...this._listeners]) {
      listener(value);
    }
  }

  async *[Symbol.asyncIterator]() {
    const queue = [];
    let pending = null;

    const unsub = this.subscribe((value) => {
      if (pending) {
        const resolve = pending;
        pending = null;
        resolve(value);
      } else {
        queue.push(value);
      }
    });

    try {
      while (true) {
        if (queue.length > 0) {
          yield queue.shift();
        } else {
          yield await new Promise((resolve) => {
            pending = resolve;
          });
        }
      }
    } finally {
      unsub();
    }
  }
}
